src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -379,288 +434,369 @@ private static HttpRequestException GetVersionException(HttpRequestMessage reque
-            }
-            finally
+            if (!_http2RequestQueue.TryPeekNextRequest(out HttpRequestMessage? request))
             {
-                _http2ConnectionCreateLock.Release();
+                return;
             }
 
-            if (sslStream != null)
+            // Determine if we can and should add a new connection to the pool.
+            if ((_availableHttp2Connections?.Count ?? 0) == 0 &&                            // No available connections
+                !_pendingHttp2Connection &&                                                 // Only allow one pending HTTP2 connection at a time
+                (_associatedHttp2ConnectionCount == 0 || EnableMultipleHttp2Connections))   // We allow multiple connections, or don't have a connection currently
             {
-                // We established an SSL connection, but the server denied our request for HTTP2.
-                // Try to establish a 1.1 connection instead and add it to the pool.
+                _associatedHttp2ConnectionCount++;
+                _pendingHttp2Connection = true;
 
-                if (NetEventSource.Log.IsEnabled())
-                {
-                    Trace("Server does not support HTTP2; disabling HTTP2 use and proceeding with HTTP/1.1 connection");
-                }
+                Task.Run(() => AddHttp2ConnectionAsync(request));
+            }
+        }
 
-                bool canUse = true;
+        private async ValueTask<Http2Connection?> GetHttp2ConnectionAsync(HttpRequestMessage request, bool async, CancellationToken cancellationToken)
+        {
+            Debug.Assert(_kind == HttpConnectionKind.Https || _kind == HttpConnectionKind.SslProxyTunnel || _kind == HttpConnectionKind.Http || _kind == HttpConnectionKind.SocksTunnel || _kind == HttpConnectionKind.SslSocksTunnel);
+
+            // Look for a usable connection.
+            TaskCompletionSourceWithCancellation<Http2Connection?> waiter;
+            while (true)
+            {
+                Http2Connection connection;
                 lock (SyncObj)
                 {
-                    // Server does not support HTTP2. Disable further HTTP2 attempts.
-                    _http2Enabled = false;
+                    _usedSinceLastCleanup = true;
 
-                    if (_associatedConnectionCount < _maxConnections)
+                    if (!_http2Enabled)
                     {
-                        IncrementConnectionCountNoLock();
+                        return null;
+                    }
+
+                    int availableConnectionCount = _availableHttp2Connections?.Count ?? 0;
+                    if (availableConnectionCount > 0)
+                    {
+                        // We have a connection that we can attempt to use.
+                        // Validate it below outside the lock, to avoid doing expensive operations while holding the lock.
+                        connection = _availableHttp2Connections![availableConnectionCount - 1];
```

Comment:
Good point. I'm also concerned by this. Currently, [GetHttp2Connection](https://github.com/dotnet/runtime/blob/1b6106596609a90e0f2db136e6186b39f34b3739/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L630) goes from the oldest to the newest connection validating their expiration. It's done to handle traffic bursts, but don't keep many connection opened longer than necessary.

Reply:
The new behavior here is different, but I don't think it's any worse than the existing behavior, and in some cases it's better.

The new behavior here is that we will use HTTP2 connections in the order that they most recently became newly available for use. So, at any given point in time, we are trying to use the most recently active connection, and only falling back to other connections if/when that connection is full and not usable.

In general, I believe this will optimize for removing unnecessary connections when a burst subsides.